### PR TITLE
[#153] Add json schema support on json inspection service

### PR DESCRIPTION
**Pull request by igarashitm**
Added inspectJsonSchema() on JsonDocumentInspectionService to build a JsonDocument from json schema input

### CODE REVIEW

File: atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        try {
+            ObjectMapper objectMapper = new ObjectMapper();
+            JsonNode rootNode = objectMapper.readTree(schema);
+            
+            JsonNode header = rootNode.get("$schema");
+            if (header == null || !header.asText().startsWith("http://json-schema.org/")) {
+                throw new JsonInspectionException(String.format("The $schema property not found or invalid: '%s'", header.asText()));
+            }
+            JsonNode type = rootNode.get("type");
+            if (type == null || !type.asText().equals("object")) {
```
Comment (rhuss) [inline atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java]: minor: probably not relevant here as `type.asText()` can not be `null` (?) but maybe a `!"object".equals(type.asText())` is safer here ?

File: atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                throw new JsonInspectionException(String.format("The property type '%s' is not supported for root node", type));
+            }
+
+            Iterator<Entry<String, JsonNode>> topFields = rootNode.get("properties").fields();
```
Comment (rhuss) [inline atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java]: NPE if there is no `properties` element, please check before (like for other fields)

File: atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(SchemaInspector.class);
+
+    public SchemaInspector() {
+    }
+
+    public JsonDocument getJsonDocument() {
+        return jsonDocument;
+    }
+
+    public void inspect(String schema) throws JsonInspectionException {
```
Comment (rhuss) [inline atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java]: What about making the inspect atomic with respect to generating the JsonDocument (would make it thread-safe, too). I.e. having a `JsonDocument inspect(String schema)` where the document itself is created within the method ?  The method then could even be `static`.

File: atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                JsonComplexType target = createJsonComplexType(entry.getKey(), entry.getValue(), null);
+                jsonDocument.getFields().getField().add(target);
+            }
+        } catch (Exception e) {
+            throw new JsonInspectionException(e);
+        }
+    }
+
+    private JsonComplexType createJsonComplexType(String name, JsonNode value, JsonComplexType parent) throws JsonInspectionException {
+        if (value.get("$ref") != null) {
```
Comment (rhuss) [inline atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java]: Is this something we do need in the future ? (i.e. does the salesforce connector use `$ref` elements in the generated JSON schema ?) // @zregvart @dhirajsb

File: atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        answer.setStatus(FieldStatus.SUPPORTED);
+        answer.setPath((parent != null ? parent.getPath() : "").concat("/").concat(name));
+        
+        switch (fieldType.asText()) {
+        case "array":
+            answer = createJsonComplexType(name, value.get("items"), parent);
+            answer.setCollectionType(CollectionType.ARRAY);
+            break;
+        case "object":
+            Iterator<Entry<String, JsonNode>> subFields = value.get("properties").fields();
```
Comment (rhuss) [inline atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java]: `properties` might be missing in the value

File: atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        
+        JsonComplexType answer = new JsonComplexType();
+        answer.setJsonFields(new JsonFields());
+        answer.setName(name);
+        answer.setStatus(FieldStatus.SUPPORTED);
+        answer.setPath((parent != null ? parent.getPath() : "").concat("/").concat(name));
+        
+        switch (fieldType.asText()) {
+        case "array":
+            answer = createJsonComplexType(name, value.get("items"), parent);
```
Comment (rhuss) [inline atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java]: `value.get("items")` might be null, but this method assumes a not-null `value` element (see above). might be good to either check here for null, or above assert that `value` is not null.

File: atlas-json-parent/atlas-json-core/src/test/resources/inspect/schema/address.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    "properties": {
+        "post-office-box": { "type": "string" },
+        "extended-address": { "type": "string" },
+        "street-address": { "type": "string" },
+        "locality":{ "type": "string" },
+        "region": { "type": "string" },
+        "postal-code": { "type": "string" },
+        "country-name": { "type": "string"}
+    },
+    "required": ["locality", "region", "country-name"],
```
Comment (rhuss) [inline atlas-json-parent/atlas-json-core/src/test/resources/inspect/schema/address.json]: I guess that `required` and `dependencies` properties are not used by the atlasmap mappings, right ?

**Review by rhuss — CHANGES_REQUESTED** [review]
thanks, lgtm !

Some minor change requests below but close to being merged.

File: atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java (hunk 4 above)
Comment (zregvart) [inline atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java]: For results containing multiple objects (think search/query) it does use `$ref`, not sure if the Jackson JSON schema support might generate those on its own.

File: atlas-json-parent/atlas-json-core/src/test/resources/inspect/schema/address.json (hunk 7 above)
Comment (igarashitm) [inline atlas-json-parent/atlas-json-core/src/test/resources/inspect/schema/address.json]: I think so, at least the InstanceInspector doesn't care if it's required nor if there's any dependency in between. Possibly we may want to add it later. BTW these schema examples come from json-schema.org official - sorry I avoided to make something optimized for the tests as I'm lazy ;)
http://json-schema.org/examples.html

I submitted a follow-up PR #155 to reflect all review comments, thanks @rhuss , @zregvart !

File: atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java (hunk 4 above)
Comment (igarashitm) [inline atlas-json-parent/atlas-json-core/src/main/java/io/atlasmap/json/inspect/SchemaInspector.java]: I'm not sure how it's displayed in atlasmap-ui right now, but it definitely affects UI as well - inline all the nested objects? just show a link? like that

**Comment by rhuss** [comment]
Oops, no ideas how I merged this PR. Strange ....

sorry ..., lets followup on #153

**Comment by rhuss** [comment]
Maybe because is used "merged" in my comment ? (though the keyword was "[merge]" ....)

**Pull request merged**